### PR TITLE
Add com.inform7.IDE

### PR DESCRIPTION
**Pull request by ddelabru**
I have been working with the maintainer of the GNOME Inform 7 IDE, @ptomato, to package [Inform 7](http://inform7.com) for Flathub. A few notes for Flathub folks:
 * The webkitgtk module is copied from the org.gimp.GIMP Flathub manifest. WebKit1 is enormous and takes several hours to compile, and the GIMP packagers had to tweak the module a lot to get it to compile. This module does not compile on org.gnome.Platform/3.30, which is why I'm targeting org.gnome.Platform/3.28 instead.
 * I'm aware that there are network security concerns regarding WebKit1, but Inform 7 does not access the network, and I have not given the Flatpak container network access. A GTK3 version of Inform 7 is in the works, but I don't know when this will see release, so for now I have to stick with WebKit1 because WebKit2 does not support GTK2.
 * The appstream XML file I included seems correct to me, but `appstream-util validate-relax` fails with errors on even the example XML from the appstream documentation on my Fedora 29 system, with errors like `/org.gnome.gnome-power-statistics.appdata.xml: failed to parse org.gnome.gnome-power-statistics.appdata.xml: Error on line 1 char 1: Document must begin with an element (e.g. <book>)`
 * While the IDE is GPL 3, the Inform 7 compiler itself, written by Graham Nelson, is not (yet) open source, though it is redistributable. So part of the manifest extracts the appropriate binary for the compiler based on the build architecture.
 * Although I've implemented certain source changes with a patch against the tarball for the most recent release of Inform 7, I've also submitted changes upstream, so packaging future releases will be easier.

### CODE REVIEW

**Comment by ddelabru** [comment]
Sorry for all the last minute `commit --amend`s, trying to tie up some minor loose ends but also start with a clean commit history. I won't tinker with it anymore until folks start reviewing/testing.

**Comment by nedrichards** [comment]
I love renaming the icon with a patch (and ideally doing it upstream) but fwiw you *could* choose to use rename-icon and then set copy-icon: true to leave the old icons behind as well so the icons continue to work in app.

**Comment by nedrichards** [comment]
bot, build com.inform7.IDE
testbot, build com.inform7.IDE

**Comment by nedrichards** [comment]
@ptomato I assume you'd like access to the repo as well once added, even if you're not going to lead the maintenance.

File: com.inform7.IDE.yml [hunk 1]
```diff
@@ -0,0 +1,94 @@
+---
+app-id: com.inform7.IDE
+runtime: org.gnome.Platform
+runtime-version: 3.28
```
Comment (bilelmoussaoui) [inline com.inform7.IDE.yml]:
```suggestion
runtime-version: '3.28'
```

**Comment by ddelabru** [comment]
@nedrichards, thanks for the heads-up; I didn't notice `copy-icon` in the options. And yes, please do give @ptomato write access as well.

Not sure what to do about flatpak-builder failing to clone the flathub/shared-modules submodule (the reason test build 1259 failed). Of course, in my local repo, I already have that cloned, so it wasn't a problem in testing.

**Comment by ddelabru** [comment]
bot, build com.inform7.IDE
testbot, build com.inform7.IDE

**Comment by nedrichards** [comment]
bot, build com.inform7.IDE

(for now only admins can use the main bot, which builds on all architectures)

**Comment by ddelabru** [comment]
Alright, test build 1260 failed for the same reason the appstream XML fails validation locally: libxml says it doesn't start with an element, but I don't understand that at all. It looks like valid XML to me?

**Comment by ddelabru** [comment]
Alright, there were some invalid invisible characters in the XML. Managed to massage it into passing `validate-relax`. Will also resubmit this XML upstream.

**Comment by ddelabru** [comment]
testbot, build com.inform7.IDE

**Comment by ddelabru** [comment]
testbot, build com.inform7.IDE

**Comment by nedrichards** [comment]
bot, build com.inform7.IDE

File: README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
+Flatpak manifest for GNOME Inform 7
+===================================
+
+This repo contains a Flatpak manifest for GNOME Inform 7, currently at version
+6M62.
+
```
Comment (bilelmoussaoui) [inline README.md]: I don't think this file is needed here.

File: com.inform7.IDE.yml [hunk 3]
```diff
@@ -0,0 +1,94 @@
+---
+app-id: com.inform7.IDE
+runtime: org.gnome.Platform
+runtime-version: '3.28'
+sdk: org.gnome.Sdk
+command: gnome-inform7
+rename-desktop-file: gnome-inform7.desktop
```
Comment (bilelmoussaoui) [inline com.inform7.IDE.yml]: It would be nice if you could rename that upstream.

File: com.inform7.IDE.yml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    path: patches/gnome-inform7.patch
+  - type: file
+    path: files/com.inform7.IDE.appdata.xml
+  - type: shell
+    commands:
+    - tar -f inform7-compilers_6M62_${INFORM7_ARCH}.tar.gz -C src/ni/ -x share/inform7/Compilers/ni --strip-components=3
+    - for dir in pixmaps/*x*; do mv ${dir}/apps/{inform7,com.inform7.IDE}.png; done
+    - mv com.inform7.{GUI,IDE}.gschema.xml
+    - mkdir /app/share/metainfo && install com.inform7.IDE.appdata.xml /app/share/metainfo/
+finish-args:
```
Comment (bilelmoussaoui) [inline com.inform7.IDE.yml]: Those should be put at the top of the manifest instead.

File: com.inform7.IDE.yml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  - type: patch
+    path: patches/gnome-inform7.patch
+  - type: file
+    path: files/com.inform7.IDE.appdata.xml
+  - type: shell
+    commands:
+    - tar -f inform7-compilers_6M62_${INFORM7_ARCH}.tar.gz -C src/ni/ -x share/inform7/Compilers/ni --strip-components=3
+    - for dir in pixmaps/*x*; do mv ${dir}/apps/{inform7,com.inform7.IDE}.png; done
+    - mv com.inform7.{GUI,IDE}.gschema.xml
+    - mkdir /app/share/metainfo && install com.inform7.IDE.appdata.xml /app/share/metainfo/
```
Comment (bilelmoussaoui) [inline com.inform7.IDE.yml]: You could use install directly here without using mkdir.

File: files/com.inform7.IDE.appdata.xml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    <p>
+      Interactive fiction is a literary form which involves programming a
+      computer so that it presents a reader with a text which can be explored.
+      Inform aims to make the burden of learning to program such texts as light
+      as possible.
+    </p>
+  ​</description>
+  <launchable type="desktop-id">com.inform7.IDE.desktop</launchable>
+  <url type="homepage">http://inform7.com/</url>
+  <provides><binary>gnome-inform7</binary></provides>
```
Comment (bilelmoussaoui) [inline files/com.inform7.IDE.appdata.xml]: Not needed; It's extracted from the desktop file.

File: patches/gnome-inform7.patch [hunk 7]
```diff
@@ -0,0 +1,671 @@
...
+@@ -1,30 +1,30 @@
+ <schemalist>
+ 
+-  <enum id="com.inform7.GUI.preferences.font-set-enum">
++  <enum id="com.inform7.IDE.preferences.font-set-enum">
+     <value nick="Standard" value="0" />
+     <value nick="Monospace" value="1" />
+     <value nick="Custom" value="2" />
+   </enum>
+
```
Comment (bilelmoussaoui) [inline patches/gnome-inform7.patch]: It would be nice if you could submit this patch upstream.

Comment (ddelabru) [reply patches/gnome-inform7.patch]: Already merged upstream. I'm only still using the patch here so I could continue using a tall of the most recent release. But future releases won't require a patch like this.

File: README.md (hunk 2 above)
Comment (ddelabru) [reply README.md]: Removed.

File: com.inform7.IDE.yml (hunk 4 above)
Comment (ddelabru) [reply com.inform7.IDE.yml]: Done.

File: files/com.inform7.IDE.appdata.xml (hunk 6 above)
Comment (ddelabru) [reply files/com.inform7.IDE.appdata.xml]: Removed.

**Comment by ddelabru** [comment]
testbot, build com.inform7.IDE

**Comment by nedrichards** [comment]
bot, build com.inform7.IDE

**Comment by ddelabru** [comment]
@nedrichards, I noticed that buildbot successfully built the app on all architectures except aarch64, where the error suggests the problem maybe an outdated `config.guess` included with goocanvas 1.0, so I've added something to the manifest to update this file.

**Comment by ddelabru** [comment]
testbot, build com.inform7.IDE

**Comment by nedrichards** [comment]
bot, build com.inform7.IDE

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/com.inform7.IDE

**Pull request closed without merging**